experimental/calendar/src/date.rs

```diff
@@ -0,0 +1,123 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::{Calendar, DateDuration, DurationUnit, Iso};
+use std::fmt;
+
+pub trait AsCalendar {
+    type Calendar: Calendar;
+    fn as_calendar(&self) -> &Self::Calendar;
+}
+
+impl<C: Calendar> AsCalendar for C {
+    type Calendar = C;
+    fn as_calendar(&self) -> &Self {
+        self
+    }
+}
+
+pub struct Date<A: AsCalendar> {
+    inner: <A::Calendar as Calendar>::DateInner,
+    calendar: A,
+}
+
+impl<A: AsCalendar> Date<A> {
+    /// Construct a date from an ISO date and some calendar representation
+    pub fn new_from_iso(iso: Date<Iso>, calendar: A) -> Self {
+        let inner = calendar.as_calendar().date_from_iso(iso);
+        Date { inner, calendar }
+    }
+
+    pub fn to_iso(&self) -> Date<Iso> {
+        self.calendar.as_calendar().date_to_iso(self.inner())
+    }
+
+    /// The number of months in the year of this date
+    pub fn months_in_year(&self) -> u8 {
+        self.calendar.as_calendar().months_in_year(&self.inner)
+    }
+
+    /// The number of days in the year of this date
+    pub fn days_in_year(&self) -> u32 {
+        self.calendar.as_calendar().days_in_year(&self.inner)
+    }
+
+    /// The number of days in the month of this date
+    pub fn days_in_month(&self) -> u8 {
+        self.calendar.as_calendar().days_in_month(&self.inner)
+    }
+
+    /// The day of the week for this date
+    ///
+    /// Monday is 1, Sunday is 7, according to ISO
+    pub fn day_of_week(&self) -> u8 {
+        self.calendar.as_calendar().day_of_week(&self.inner)
+    }
+
+    /// Add a `duration` to this date, mutating it
+    pub fn add(&mut self, duration: DateDuration<A::Calendar>) {
+        self.calendar
+            .as_calendar()
+            .offset_date(&mut self.inner, duration)
+    }
+
+    /// Calculating the duration between `other - self`
+    pub fn until<B: AsCalendar<Calendar = A::Calendar>>(
+        &self,
+        other: &Date<B>,
+        largest_unit: DurationUnit,
+        smallest_unit: DurationUnit,
+    ) -> DateDuration<A::Calendar> {
+        self.calendar
+            .as_calendar()
+            .until(self.inner(), other.inner(), largest_unit, smallest_unit)
+    }
+
+    /// Construct a date from raw values for a given calendar. This does not check any
+    /// invariants for the date and calendar, and should only be called by calendar implementations.
+    ///
+    /// Calling this outside of calendar implementations is sound, but calendar implementations are not
+    /// expected to do anything sensible with such invalid dates.
+    pub fn construct_unchecked(inner: <A::Calendar as Calendar>::DateInner, calendar: A) -> Self {
+        Self { inner, calendar }
+    }
```

Comment:
Thought: Hmm. I guess my expectation here is that `DateInner` is the thing that needs to be valid. `Date` itself doesn't care at all about what's in `DateInner`. So perhaps this method does not need to be named "unchecked", since it gets passed a DateInner, which should have already been validated.

Reply:
a particular DateInner may only be valid for a particular _instance_ of C, e.g. dates from non-matching tabular islamic calendars.

Reply:
Maybe we can just architect the code so that this method does not need to be `pub`.

You have two call sites:

```rust
impl Calendar for Iso {
    fn date_to_iso(&self, date: &Self::DateInner) -> Date<Iso> {
        Date::construct_unchecked(*date, Iso)
    }
}
```

This could be resolved by either:

1. Making a special function on `Date<Iso>` called `new_from_inner()` that takes a `IsoDateInner` as an argument (which should always succeed since there is only one `Iso` calendar)
2. Make the trait function `date_to_iso` take a `IsoDateInner` as an argument, instead of `Date<Iso>`, as suggested in one of my other comments

Reply:
The other call site:

```rust
impl Iso {
    pub fn create_iso_date(day: IsoDay, month: IsoMonth, year: IsoYear) -> Result<Date<Iso>, ()> {
        if day.0 > 28 {
            let bound = Self::days_in_month(year, month);
            if day.0 < bound {
                return Err(());
            }
        }

        Ok(Date::construct_unchecked(
            IsoDateInner { day, month, year },
            Iso,
        ))
    }
```

Seems like this function doesn't belong on `Iso` anyway; it should go on `Date<Iso>`.

Reply:
> You have two call sites:

There will be more call sites once we support more calendars, and they will not be in the same crate.

Reply:
Two parts of this response.

## The case to remove the method

The model I'd like is if `Date` can call methods in the `Calendar` trait that return a `DateInner`.

I listed the two call sites where you need this.

1. In the `Calendar::date_to_iso` function, we can work around this fairly easily by changing the signature in a number of ways.
2. The `create_xxx_date` function is the more interesting case.  Perhaps we could add (yet another) associated type called something like `DateInput`, and then create a function like:

```rust
impl Date<A> where A: AsCalendar<C = Calendar> {
    fn from_fields(fields: A::Calendar::DateInput, calendar: A) {
        let validated_fields: A::Calendar::DateInner = calendar.validate(fields);
        Self { inner: validated_fields, calendar }
    }
}
```

But maybe that is ergonomically awkward and constraining for calendar implementors.

## The case to keep the method

Acknowledged about cross-crate.  My suggestion was intended to remove the need for cross-crate, but maybe there are still gaps.

Maybe if we renamed the method, I would like it better.  How about `from_raw`?  To me `unchecked` means that something is `unsafe`, whereas here, calling this method won't put your application at risk; it just might trigger unexpected behavior in some odd edge cases like mismatched tabular calendars.

Reply:
> But maybe that is ergonomically awkward and constraining for calendar implementors.

Yeah, it's constraining, most implementations will be constructing it, and it's redundant to validate it later.

I'm fine with naming it as from_raw!